src/reconcile/rec_visibility.c

```diff
@@ -245,11 +249,19 @@ __rec_find_and_save_delete_hs_upd(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_
                 break;
             }
         }
+
+        /* Track the first globally visible self-contained value. */
+        if (!F_ISSET(r, WT_REC_EVICT) && visible_all_upd == NULL &&
+          __wt_txn_upd_visible_all(session, delete_upd) && WT_UPDATE_DATA_VALUE(delete_upd))
+            visible_all_upd = delete_upd;
     }
 
     WT_ASSERT_ALWAYS(session, delete_tombstone == NULL || delete_upd != NULL,
       "If we delete a tombstone from the history store, we must also delete the update.");
 
+    /* Free obsolete updates if exist. */
+    if (!delete_hs_upd_found && visible_all_upd != NULL && visible_all_upd->next != NULL)
```

Comment:
Why we need to call obsolete check both in reconciliation and hs code? I think one place is enough?

Reply:
As the oldest timestamp moves in parallel to the reconciliation, it would be better to do the cleanup in both reconciliation and hs flows. The cost to free is much less if we have an obsolete update than retaining it.

Reply:
Long checkpoint is already a problem so I'm worried it will exacerbate the problem. I'm OK to do this way now as I don't have any data.

Reply:
I agree with you that it may increase the checkpoint duration. But we call the obsolete check function only when we find a globally visible update, so the cost of freeing the remaining update list may not be that overhead.